### ISSUE TITLE
per-shard readyness metrics

### ISSUE THREAD

**Issue by Dieterbe** [grafana/metrictank#1492]
needs a bit more thought, but basically dictated by new deployment model. we used to check number of replicas of deployments, but now we're deploying multiple copies of individual pods (with k8s replication=1). need a better way to monitor readyness of shards

**Comment by replay** [comment]
considering that in the future those deployment models might become a bit more complex (IIRC we talked about partially overlapping shard groups), it might be better to monitor the readiness per partition instead of per shard group.

**Comment by woodsaj** [comment]
There is already a metric for the kafka lag, but this is the number of messages behind that we are.

`metrictank.stats.$cluster.mt-read*.input.kafka-mdm.partition.*.lag.gauge64`

What we really need is for https://github.com/bloomberg/metrictank/blob/master/input/kafkamdm/lag_monitor.go to export the per partition 'priority' numbers as a metric.  This wouldnt' be too hard.

we can then just do a simple groupBy to get the lowest "priority" per partition Id and alert if it is too high.


* as a note, i would also like to have the "clustering" package store the per partition "priority" numbers instead of only getting the max priority.  ie,  each cluster node should tell its peers what partitions it has, and for each of those partitions, what the "priority" is.  This will support future cluster topology changes we have discussed.

**Comment by replay** [comment]
There are still possible, although unlikely, cases where we can't detect that a partition is unavailable. For example in a situation where 2 pods have the same 2 partitions with these priorities:

```
mt-read00:
0: 1000
1: 1

mt-read01
0: 1
1: 10000
```

Both nodes would be seen as non-ready, even though each partition has one ready node.

**Comment by Dieterbe** [comment]
Per shard priority seems useful for debugging , but per shard readiness seems more useful for high level monitoring whether the shard (=partition) is actually ready.

**Comment by replay** [comment]
I think to support future use cases and deployment models it would be better if we would change the ready-status to be per partition, instead of it being per Metrictank.
As a first step we could report per-partition readiness metrics, although at the moment all partitions of a Metrictank would always become ready/non-ready together. Using these metrics we could then solve the above problem. In a next step we'd make cluster members broadcast their readiness status per partition, instead of per node.

**Comment by Dieterbe** [comment]
yep that's also what i was thinking.

**Comment by woodsaj** [comment]
> to support future use cases and deployment models it would be better if we would change the ready-status to be per partition, instead of it being per Metrictank.

We need to be careful here about what we are talking about when referring to "ready-status".  There are a few types of "ready" that we need to support.

1) partition readiness:  This is based purely on whether this partition is up-to-date or lagging behind what is in Kafka.
2 )Node readiness.  We use this in k8s deployments so we can signal to the k8s scheduler when a node has completed backfilling data for all partitions.  A MT pod shouldnt be marked ready until all partitions are ready.  ie if we have 2 replicas, we need to ensure that all partitions on replica1 are ready before we kill replica2.

**Comment by replay** [comment]
@fkaleo made a good suggestion earlier, that the partition readiness metric should only say a partition is ready once its partition is <10 AND the node is also ready. that way its easy to create a metric that alerts us if a partition's data is not served by any cluster node.
Although, if the node is only ready once all its partitions are ready, then we might as well just say that a partition is ready if its node is ready. 
So the plan still stays the same for now:
- we're going to create a per-partition readiness metric which is saying a partition is ready if the node is ready (which implies the partition priority is <10)
- then we can easily alert based on that if a partition has no ready node

**Comment by Dieterbe** [comment]
The PR doesn't seem like the proper solution. (though it is very simple so maybe we should merge it anyway, and then redo the metrics again later)

Here's how i think about the proper solution (which will take more work):
note that currently (in master), communication happens like so:
`kafka input plugin -> cluster.Manager.SetPartitions() -> cluster.Manager`
`kafka input plugin -> cluster.Manager.SetPriority(k.lagMonitor.Metric()) > cluster.Manager`
where the lagMonitor.Metric is simply the worst priority value seen across all its partitions.
note also that the lagMonitor has an "explanation" struct and endpoint which contains per-partition lag, rate and priority and overall priority. this gets called like so:
`/priority -> Server.explainPriority() -> PrioritySetter.ExplainPriority()` where the main cmd has registered the kafka plugin as a prioritysetter at startup.

As we are starting to transition to partition-granularity clustering, it makes sense to try to track all partition properties on the cluster node itself (rather then having the input plugin query for clustering information, which seems backwards, albeit very pragmatic)

My proposal is:
* track per-partition state on the Node (HTTPNode) level:
```
...
type HTTPNode struct {
       Partitions     []int32                `json:"partitions"`
+       PartitionState map[int]PartitionState `json:"partitionState"`
       ApiPort        int                    `json:"apiPort"`
...
+ type PartitionState struct {
+         Lag      int // number of metrics behind (note: not always known. depends on input plugin)
+         Rate     int // ingestion rate
 +        Priority int // "score" of partition. Lower is better. for queue based input plugins: estimate of lag + seconds behind (derived from rate and lag), for carbon: always 0.
+         Ready    bool
+ }
```

* remove all PrioritySetter/explainpriority stuff. replace the cluster manager `SetPriority()` method with a new method that will report all partition status information to the cluster manager. so the kafka input plugin would still do the computations of the priority (using the heuristic we have that looks at the rate and lag numbers), but will report those rate and lag numbers, as well as the computed priority to the cluster manager. (carbon plugin will also report the priority as it does now, but not a lag number)
the cluster manager then takes all that information, tracks it on the local node level and uses it to compute the per-partition and per-node readiness. It gossips all of these as well (but we don't use the gossipped per-partition state yet) and it sends out the new metrics we need, along with any other per-partition metrics we may want.

I also envision some future next steps that will take a bit more consideration (e.g. requires the above changes to have been deployed before we can proceed):
* start using the gossipped per-partition information (and readiness) for clustering
* possibly remove node-level priority value

**Comment by replay** [comment]
@Dieterbe I agree with the plan you described, that all makes sense. This current PR is only supposed to be a quick fix to get our alerting back into a more useful state.
